File: ruby/bob/bob.rb
class Bob

  def hey(message)
    msg = Message.new(message)
    return "Fine. Be that way!" if msg.silent?
    return "Woah, chill out!" if msg.yelling?
    return "Sure." if msg.question?
    "Whatever."
  end

end

class Message

  def initialize(message)
    @message = message || ""
  end

  def question?
    @message.end_with? "?"
  end  

  def yelling?
    @message.upcase == @message
  end

  def silent?
    @message.strip == ""
  end

end

File: elixir/bob/bob.exs
defprotocol MessageType do
  def question?(message)
  def yelling?(message)
  def silent?(message)
end

defimpl MessageType, for: BitString do
  def question?(message) do
    String.ends_with? message, "?"
  end

  def yelling?(message) do
    String.upcase(message) == message
  end

  def silent?(message) do
    String.strip(message) == ""
  end
end

defmodule Teenager do

  def hey(message) do
    respond identify_message(message)
  end

  def respond({:question}), do: "Sure."
  def respond({:silent}), do: "Fine. Be that way!"
  def respond({:yelling}), do: "Woah, chill out!"
  def respond(_), do: "Whatever."

  defp identify_message(message) do
    cond do
      MessageType.question?(message) ->
        {:question}
      MessageType.silent?(message) ->
        {:silent}
      MessageType.yelling?(message) ->
        {:yelling}
      true ->
        {:unidentified}
    end
  end

end